src/main/java/de/tum/in/www1/artemis/service/competency/CompetencyProgressService.java

```diff
@@ -217,26 +204,187 @@ public CompetencyProgress updateCompetencyProgress(Long competencyId, User user)
      *
-     * @param learningObjects A list of all learning objects linked to a specific competency
-     * @param user            The user for which the progress should be calculated
-     * @return The percentage of completed learning objects by the user
+     * @param exerciseInfos      The information about the exercises linked to the competency
+     * @param competencyProgress The progress entity to update
      */
-    private double calculateProgress(@NotNull Set<LearningObject> learningObjects, @NotNull User user) {
-        return learningObjects.stream().map(learningObject -> learningObjectService.isCompletedByUser(learningObject, user)).mapToInt(completed -> completed ? 100 : 0).average()
-                .orElse(0.);
+    private void calculateConfidence(Set<CompetencyExerciseMasteryCalculationDTO> exerciseInfos, CompetencyProgress competencyProgress) {
+        Set<CompetencyExerciseMasteryCalculationDTO> participantScoreInfos = exerciseInfos.stream()
+                .filter(info -> info.lastScore() != null && info.lastPoints() != null && info.lastModifiedDate() != null).collect(Collectors.toSet());
+
+        double recencyConfidenceHeuristic = calculateRecencyConfidenceHeuristic(participantScoreInfos);
+        double difficultyConfidenceHeuristic = calculateDifficultyConfidenceHeuristic(participantScoreInfos, exerciseInfos);
+        double quickSolveConfidenceHeuristic = calculateQuickSolveConfidenceHeuristic(participantScoreInfos);
+
+        double confidence = 1 + recencyConfidenceHeuristic + difficultyConfidenceHeuristic + quickSolveConfidenceHeuristic;
+
+        competencyProgress.setConfidence(confidence);
+        setConfidenceReason(competencyProgress, recencyConfidenceHeuristic, difficultyConfidenceHeuristic, quickSolveConfidenceHeuristic);
     }
 
     /**
-     * Calculate the confidence score for the given user in a competency.
+     * Calculate the recency confidence heuristic for the given user in a competency based on the exercises linked to the competency.
+     * If the recent scores are higher than the average scores, the confidence should also be higher and vice versa.
      *
-     * @param exercises A list of all exercises linked to a specific competency
-     * @param user      The user for which the confidence score should be calculated
-     * @return The average score of the user in all exercises linked to the competency
+     * @param participantScores the participant scores for the exercises linked to the competency
+     * @return The recency confidence heuristic
      */
-    private double calculateConfidence(@NotNull Set<Exercise> exercises, @NotNull User user) {
-        return participantScoreService.getStudentAndTeamParticipationScoresAsDoubleStream(user, exercises).summaryStatistics().getAverage();
+    private double calculateRecencyConfidenceHeuristic(@NotNull Set<CompetencyExerciseMasteryCalculationDTO> participantScores) {
+        if (participantScores.size() < 3) {
```

Comment:
Could we give the 3 a name to explicitly state what we want to express here?

As I understand it 3 is the "MINIMUM_SCORES_FOR_RECENCY_CONFIDENCE_HEURISTIC" ?